### DISCUSSION TITLE
On ui-nav-stack: custom UINavigationControllerDelegate

### DISCUSSION THREAD

**Discussion by notcome** [pointfreeco/swift-composable-architecture#3012] [Beta]
I came across the `ui-nav-stack` branch in a discussion post and found it to be working perfectly. However, I noticed that `StackNavigationController` is its own `UINavigationControllerDelegate`. As mentioned in your comment, this is done to sync UIKit view state back to TCA state.

The reason I cannot switch to NavigationStack is that I need to be able to interact with the lifecycle of UINavigationController for custom and interruptible transitions. Therefore, we need to be able to set our own delegate for UINavigationController, which is at odds with the current implementation. While we could manually hijack the delegate to forward the implementation, it would require us to closely monitor your implementation. Since this branch is still in preview, I just want to make you aware of my use case. Thank you!

**Comment by acosmicflamingo** [comment]
To add, it could perhaps be helpful for all the stack support to reside in a protocol that any custom navigation controller can conform to, to make it very easy for us to either use the TCA-provided navigation controller or use our own custom navigation controller with that functionality.

**Comment by mbrandonw** [comment]
> The reason I cannot switch to NavigationStack is that I need to be able to interact with the lifecycle of UINavigationController for custom and interruptible transitions. Therefore, we need to be able to set our own delegate for UINavigationController, which is at odds with the current implementation.

Hi @notcome, thanks for bringing this up. This is definitely something we should support. Probably the easiest way to do it is for us to have our own `StackNavigationControllerDelegate` that inherits from `UINavigationControllerDelegate` so that we can internally be the delegate of the nav controller, and externally others can become the delegate of our controller.
